Add jump to error location from the error message (in stacktrace) - Addresses #715.

turbopape opened the pull request:
If you hit enter or click with button 2 on the errors cause (present at error-buffer) which have informatin about line/column of the causes, you are automatically taken to the file and the exact location of that cause.
This addresses #715.

turbopape commented:
@bbatsov Have you had the time to look into this yet? Do I need to change anything in order to make this PR conform to CIDER's standards ?

Malabarba commented:
Hi there. Sorry noone looked into it yet. Could you start by fixing the way `let` forms are formatted? 
Also make sure you select that bit of code and hit TAB. There  are a couple of places where indentation seems a bit off.

turbopape commented:
Thank you for your response. I've fixed the indentation and let forms formatting as per your recommendations !

expez commented on cider-stacktrace.el:
```diff
@@ -518,23 +518,51 @@ This associates text properties to enable filtering and source navigation."
-            (when data
-              (cider-stacktrace-emit-indented
-               (cider-font-lock-as-clojure data) indent nil)))
+                                   (let ((beg (point)))
```
You have 3 nested `let`s here.  As far as I can tell you only need one.

expez commented on cider-stacktrace.el:
```diff
@@ -518,23 +518,51 @@ This associates text properties to enable filtering and source navigation."
+                                     (cider-stacktrace-emit-indented
+                                      (propertize (or message "(No message)") 'font-lock-face  message-face) indent t)
+                                     
+                                     (let* ((toks (split-string message ":")))
```
This function does too much now.  I think a natural split point is to create something like a `create-go-to-err-action` function which creates the action used in the button.  If you think it's necessary that function might even delegate to another function which picks apart the message.

Malabarba replied:
This let form and the ones below are very hard to read. Here's what you could use instead:

    (when (string-match "\\([^:]+\\):\\([^:]+\\):\\([^:]+\\)\\'" message)
      (let ((line (string-to-number (match-string 3 message)))
            (col (string-to-number (match-string 2 message)))
            (filename (match-string 1 message))))
      ...)

turbopape replied:
Ok guys,

I'll have a look into this If I manage some time in the afternoon...

Thank  you !

2015-09-11 11:04 GMT+01:00 Artur Malabarba <notifications@github.com>:

> In cider-stacktrace.el
> <https://github.com/clojure-emacs/cider/pull/1309#discussion_r39257473>:
>
> > @@ -518,23 +518,51 @@ This associates text properties to enable filtering and source navigation."
> >              (newline))
> >            ;; Detail level 1: message + ex-data
> >            (cider-propertize-region '(detail 1)
> > -            (cider-stacktrace-emit-indented
> > -             (propertize (or message "(No message)") 'font-lock-face message-face) indent t)
> > -            (newline)
> > -            (when data
> > -              (cider-stacktrace-emit-indented
> > -               (cider-font-lock-as-clojure data) indent nil)))
> > +                                   (let ((beg (point)))
> > +                                     (cider-stacktrace-emit-indented
> > +                                      (propertize (or message "(No message)") 'font-lock-face  message-face) indent t)
> > +
> > +                                     (let* ((toks (split-string message ":")))
>
> This let form and the ones below are very hard to read. Here's what you
> could use instead:
>
> (when (string-match "\\([^:]+\\):\\([^:]+\\):\\([^:]+\\)\\'" message)
>   (let ((line (string-to-number (match-string 3 message)))
>         (col (string-to-number (match-string 2 message)))
>         (filename (match-string 1 message))))
>   ...)
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/clojure-emacs/cider/pull/1309/files#r39257473>.
>

Malabarba commented:
>  I've fixed the indentation 

Sorry, there was problem in the definition of `cider-propertize-region` that caused your indentation to be all off now. This isn't your fault.
I've fixed the definition of `cider-propertize-region`  on master. Could you rebase to the current master, reevaluate `cider-util.el` and then reindent your whole code again?

> and let forms formatting as per your recommendations !

There's still one `let*` that looks wrong in there. Though maybe you could just merge the `let`s together like expez suggests.

turbopape commented:
Hi Guys, Can you have a look and tell me what it looks like now ?

expez commented on cider-stacktrace.el:
```diff
@@ -518,8 +533,11 @@ This associates text properties to enable filtering and source navigation."
+            (let ((beg (point)))
+              (cider-stacktrace-emit-indented
+               (propertize (or message "(No message)") 'font-lock-face  message-face) indent t)
+              (print "hello!!")
```
This probably shouldn't be here

expez commented on cider-stacktrace.el:
```diff
@@ -502,6 +502,21 @@ This associates text properties to enable filtering and source navigation."
             (put-text-property p2 p3 'font-lock-face 'cider-stacktrace-fn-face)))
         (newline)))))
 
+(defun create-go-to-err-action (beg message)
```
This function has to be namespaced:  `cider-stacktrace--create-go-to-err-action`

expez commented on cider-stacktrace.el:
```diff
@@ -502,6 +502,21 @@ This associates text properties to enable filtering and source navigation."
+           (col (string-to-number (match-string 4 message)))
+           (buf-name (car (last (split-string (match-string 2 message) "\\/"))))
+           (the-buf-window (get-buffer-window buf-name)))
+      (make-button (+ beg 3)
```
What happens if you replace `beg` with `(point)`?  I might be misreading the code but I looks like it would give the same result.

turbopape replied:
I use the let beg... stuff to store the point  before the message is
emitted. Then in the function I use make-button on the region from beg to
current (point). Is there a better way to do it?
Le 11 sept. 2015 12:52, "Lars Andersen" <notifications@github.com> a écrit :

> In cider-stacktrace.el
> <https://github.com/clojure-emacs/cider/pull/1309#discussion_r39263618>:
>
> > @@ -502,6 +502,21 @@ This associates text properties to enable filtering and source navigation."
> >              (put-text-property p2 p3 'font-lock-face 'cider-stacktrace-fn-face)))
> >          (newline)))))
> >
> > +(defun create-go-to-err-action (beg message)
> > +  (when (string-match "\\([^:]+\\):\\([^:]+\\):\\([^:]+\\):\\([^:]+\\)\\'" message)
> > +    (let* ((line (string-to-number (match-string 3 message)))
> > +           (col (string-to-number (match-string 4 message)))
> > +           (buf-name (car (last (split-string (match-string 2 message) "\\/"))))
> > +           (the-buf-window (get-buffer-window buf-name)))
> > +      (make-button (+ beg 3)
>
> What happens if you replace beg with (point)? I might be misreading the
> code but I looks like it would give the same result.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/clojure-emacs/cider/pull/1309/files#r39263618>.
>

expez replied:
No, I think you're right.  Sorry about the noise.

expez commented on cider-stacktrace.el:
```diff
@@ -502,6 +502,21 @@ This associates text properties to enable filtering and source navigation."
         (newline)))))
 
+(defun create-go-to-err-action (beg message)
+  (when (string-match "\\([^:]+\\):\\([^:]+\\):\\([^:]+\\):\\([^:]+\\)\\'" message)
```
I wonder if there's something we can do here in the event that the message doesn't match the expected format.  As it stands we'll get a crash when we try to use `nil` as an action in the button.  Perhaps this should be an `if` with an error in the other branch?

turbopape commented:
I namespaced the go-to-err-action function.
Besides, I added an if condition so we are sure not to create a button if the buffer name is nil.
As per the first format pattern matching, nothing will happen as the when block protects all of the function body.
Tell me what you think of this commit ...

expez commented on cider-stacktrace.el:
```diff
@@ -502,6 +502,23 @@ This associates text properties to enable filtering and source navigation."
+          (make-button (+ beg 3)
+                       (point)
+                       'action (lambda (button)
+                                 (if the-buf-window
```
You now have two nested ifs testing for the same thing.

Malabarba replied:
Also, the `if` inside the lambda is unreliable. The window might no longer exist, or it might displaying another buffer by the time this lambda is called.
Getting the window for the buffer should happen inside the lambda.

turbopape replied:
Okay I get it. I'll change this.


2015-09-11 17:31 GMT+01:00 Artur Malabarba <notifications@github.com>:

> In cider-stacktrace.el
> <https://github.com/clojure-emacs/cider/pull/1309#discussion_r39289273>:
>
> > @@ -502,6 +502,23 @@ This associates text properties to enable filtering and source navigation."
> >              (put-text-property p2 p3 'font-lock-face 'cider-stacktrace-fn-face)))
> >          (newline)))))
> >
> > +(defun cider-stacktrace--create-go-to-err-action (beg message)
> > +  (when (string-match "\\([^:]+\\):\\([^:]+\\):\\([^:]+\\):\\([^:]+\\)\\'" message)
> > +    (let* ((line (string-to-number (match-string 3 message)))
> > +           (col (string-to-number (match-string 4 message)))
> > +           (buf-name (car (last (split-string (match-string 2 message) "\\/"))))
> > +           (the-buf-window (if buf-name
> > +                               (get-buffer-window buf-name))))
> > +      (if the-buf-window
> > +          (make-button (+ beg 3)
> > +                       (point)
> > +                       'action (lambda (button)
> > +                                 (if the-buf-window
>
> Also, the if inside the lambda is unreliable. The window might no longer
> exist, or it might displaying another buffer by the time this lambda is
> called.
> Getting the window for the buffer should happen inside the lambda.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/clojure-emacs/cider/pull/1309/files#r39289273>.
>

expez commented on cider-stacktrace.el:
```diff
@@ -502,6 +502,23 @@ This associates text properties to enable filtering and source navigation."
+           (buf-name (car (last (split-string (match-string 2 message) "\\/"))))
+           (the-buf-window (if buf-name
+                               (get-buffer-window buf-name))))
+      (if the-buf-window
```
It's customary to use `when` when the `if` clause doesn't have two branches

turbopape commented:
I changed my single branch ifs with whens.
Regarding the two ifs, actually they are not testing for the same thing. buf-name could be not null, yet there is no window with that buffer name, so window-buf-name could be null and then we could not get there and jump to the error. These two tests are not the same.

turbopape commented:
I get buffer-window inside the lambda now. I test if it is found else I open the buffer with switch buffer. Is it right ?

bbatsov commented on cider-stacktrace.el:
```diff
@@ -502,6 +502,22 @@ This associates text properties to enable filtering and source navigation."
             (put-text-property p2 p3 'font-lock-face 'cider-stacktrace-fn-face)))
         (newline)))))
 
+(defun cider-stacktrace--create-go-to-err-action (beg message)
```
Even though this is private, a docstring would still be useful.

bbatsov replied:
I'm guessing that in the name it'd be better to replace "action" with "button".

expez replied:
The button isn't created here, though, but down on L511.

bbatsov replied:
Exactly. It's the output of this function, so I think the name is confusing.

bbatsov commented on cider-stacktrace.el:
```diff
@@ -502,6 +502,22 @@ This associates text properties to enable filtering and source navigation."
+    (let* ((line (string-to-number (match-string 3 message)))
+           (col (string-to-number (match-string 4 message)))
+           (buf-name (car (last (split-string (match-string 2 message) "\\/")))))
+      (when buf-name
```
The indentation is incorrect here.

Malabarba replied:
Is it?

turbopape replied:
Oops! Didn't I save???
Le 13 sept. 2015 12:06, "Artur Malabarba" <notifications@github.com> a
écrit :

> In cider-stacktrace.el
> <https://github.com/clojure-emacs/cider/pull/1309#discussion_r39346647>:
>
> > @@ -502,6 +502,22 @@ This associates text properties to enable filtering and source navigation."
> >              (put-text-property p2 p3 'font-lock-face 'cider-stacktrace-fn-face)))
> >          (newline)))))
> >
> > +(defun cider-stacktrace--create-go-to-err-action (beg message)
> > +  (when (string-match "\\([^:]+\\):\\([^:]+\\):\\([^:]+\\):\\([^:]+\\)\\'" message)
> > +    (let* ((line (string-to-number (match-string 3 message)))
> > +           (col (string-to-number (match-string 4 message)))
> > +           (buf-name (car (last (split-string (match-string 2 message) "\\/")))))
> > +      (when buf-name
>
> Is it?
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/clojure-emacs/cider/pull/1309/files#r39346647>.
>

bbatsov commented:
Guess you should also refer to this ticket https://github.com/clojure-emacs/cider/issues/715 in your commit message

turbopape commented:
Here you are ! Done indenting according to @Malabarba new formatting.

turbopape commented:
What about this commit @Malabarba ? cc @bbatsov .

Malabarba commented:
Looks good to me.

expez commented:
I tried to test this the other day but I'm not sure how to trigger this code path.  Care to enlighten me @turbopape?

turbopape commented:
Sure. You load any clj file with a deliberate error in it. When the
stacktrace pops up, you will be able to click on any error message
containing a line:col reference, which will take you to this location on
the corresponding window.
Le 13 sept. 2015 17:43, "Lars Andersen" <notifications@github.com> a écrit :

> I tried to test this the other day but I'm not sure how to trigger this
> code path. Care to enlighten me @turbopape <https://github.com/turbopape>?
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/clojure-emacs/cider/pull/1309#issuecomment-139893637>.
>

bbatsov commented:
The commits have to squashed together. The issue has to be referred in the
commit message.

On Sunday, September 13, 2015, Rafik NACCACHE <notifications@github.com>
wrote:

> Sure. You load any clj file with a deliberate error in it. When the
> stacktrace pops up, you will be able to click on any error message
> containing a line:col reference, which will take you to this location on
> the corresponding window.
> Le 13 sept. 2015 17:43, "Lars Andersen" <notifications@github.com
> <javascript:_e(%7B%7D,'cvml','notifications@github.com');>> a écrit :
>
> > I tried to test this the other day but I'm not sure how to trigger this
> > code path. Care to enlighten me @turbopape <https://github.com/turbopape
> >?
> >
> > —
> > Reply to this email directly or view it on GitHub
> > <https://github.com/clojure-emacs/cider/pull/1309#issuecomment-139893637
> >.
> >
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/clojure-emacs/cider/pull/1309#issuecomment-139893812>.
>


-- 
Best Regards,
Bozhidar Batsov

http://www.batsov.com

turbopape commented:
Ok. I'll close this PR, squash my commits and then open a new one.
Le 13 sept. 2015 17:48, "Bozhidar Batsov" <notifications@github.com> a
écrit :

> The commits have to squashed together. The issue has to be referred in the
> commit message.
>
> On Sunday, September 13, 2015, Rafik NACCACHE <notifications@github.com>
> wrote:
>
> > Sure. You load any clj file with a deliberate error in it. When the
> > stacktrace pops up, you will be able to click on any error message
> > containing a line:col reference, which will take you to this location on
> > the corresponding window.
> > Le 13 sept. 2015 17:43, "Lars Andersen" <notifications@github.com
> > <javascript:_e(%7B%7D,'cvml','notifications@github.com');>> a écrit :
> >
> > > I tried to test this the other day but I'm not sure how to trigger this
> > > code path. Care to enlighten me @turbopape <
> https://github.com/turbopape
> > >?
> > >
> > > —
> > > Reply to this email directly or view it on GitHub
> > > <
> https://github.com/clojure-emacs/cider/pull/1309#issuecomment-139893637
> > >.
> > >
> >
> > —
> > Reply to this email directly or view it on GitHub
> > <https://github.com/clojure-emacs/cider/pull/1309#issuecomment-139893812
> >.
> >
>
>
> --
> Best Regards,
> Bozhidar Batsov
>
> http://www.batsov.com
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/clojure-emacs/cider/pull/1309#issuecomment-139893904>.
>

expez commented:
@turbopape you can squash and force push to the branch to save yourself some keystrokes.

turbopape commented:
I squashed the commits. Can you verify @bbatsov @Malabarba please ?

expez commented:
@turbopape drop the trailing `.` in the commit message and use `[Fix #715]` as a prefix instead of `Related to Issue #715:`, that way it matches the conventions in the [contribution guidelines](https://github.com/clojure-emacs/cider/blob/master/CONTRIBUTING.md).

Thanks for seeing this through, we've gone quite a few rounds back and forth now!

expez commented on cider-stacktrace.el:
```diff
@@ -502,39 +506,61 @@ This associates text properties to enable filtering and source navigation."
-          ;; Add line break between causes, even when collapsed.
-          (cider-propertize-region '(detail 0)
-            (newline)))))))
+                          (let ((indent "   ")
```
Am I misreading this diff or is this `let` indented way too much to the right?

turbopape replied:
Nope this let comes just under the upper newline.
Le 13 sept. 2015 21:46, "Lars Andersen" <notifications@github.com> a écrit :

> In cider-stacktrace.el
> <https://github.com/clojure-emacs/cider/pull/1309#discussion_r39351853>:
>
> > -            (newline)
> > -            (when data
> > -              (cider-stacktrace-emit-indented
> > -               (cider-font-lock-as-clojure data) indent nil)))
> > -          ;; Detail level 2: stacktrace
> > -          (cider-propertize-region '(detail 2)
> > -            (newline)
> > -            (let ((beg (point))
> > -                  (bg `(:background ,cider-stacktrace-frames-background-color)))
> > -              (dolist (frame stacktrace)
> > -                (cider-stacktrace-render-frame buffer frame))
> > -              (overlay-put (make-overlay beg (point)) 'font-lock-face bg)))
> > -          ;; Add line break between causes, even when collapsed.
> > -          (cider-propertize-region '(detail 0)
> > -            (newline)))))))
> > +                          (let ((indent "   ")
>
> Am I misreading this diff or is this let indented way too much to the
> right?
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/clojure-emacs/cider/pull/1309/files#r39351853>.
>

turbopape replied:
No this is indented right just under the "cause" in the line just before the deletions in the diff.

expez replied:
If you look at how the code was before, you'll see that it's not supposed to be indented that far in.  You really shouldn't be indenting code manually, emacs has pretty good facilities for taking such work off your hands.  

The reason the indentation is different, from regular function indentations where the arguments are indeed aligned like you've done here, is that `nrepl-dbind-response` is a macro which sets up a context (much like `with-open-file` in clojure).

turbopape replied:
Actually I did not indent anything by hand ! I tabbed all the way in !
Anyway, I reloaded cider-util.el and re-tabbed. tell me if it is okay now ?
cc @malabarba @bbatsov

2015-09-14 11:53 GMT+01:00 Lars Andersen <notifications@github.com>:

> In cider-stacktrace.el
> <https://github.com/clojure-emacs/cider/pull/1309#discussion_r39381249>:
>
> > -            (newline)
> > -            (when data
> > -              (cider-stacktrace-emit-indented
> > -               (cider-font-lock-as-clojure data) indent nil)))
> > -          ;; Detail level 2: stacktrace
> > -          (cider-propertize-region '(detail 2)
> > -            (newline)
> > -            (let ((beg (point))
> > -                  (bg `(:background ,cider-stacktrace-frames-background-color)))
> > -              (dolist (frame stacktrace)
> > -                (cider-stacktrace-render-frame buffer frame))
> > -              (overlay-put (make-overlay beg (point)) 'font-lock-face bg)))
> > -          ;; Add line break between causes, even when collapsed.
> > -          (cider-propertize-region '(detail 0)
> > -            (newline)))))))
> > +                          (let ((indent "   ")
>
> If you look at how the code was before, you'll see that it's not supposed
> to be indented that far in. You really shouldn't be indenting code
> manually, emacs has pretty good facilities for taking such work off your
> hands.
>
> The reason the indentation is different, from regular function
> indentations where the arguments are indeed aligned like you've done here,
> is that nrepl-dbind-response is a macro which sets up a context (much
> like with-open-file in clojure).
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/clojure-emacs/cider/pull/1309/files#r39381249>.
>

turbopape commented:
Done @expez. cc @Malabarba @bbatsov .

turbopape commented:
Any thoughts about this PR @expez @Malabarba @bbatsov ?

turbopape commented:
And what do you think now @expez @bbatsov @Malabarba ?

expez commented:
Except for a missing space between `[Fix #715]` and the rest of the commit header this looks good to me.  I'll leave it to @bbatsov to merge this when he has a spare moment.

Malabarba commented on cider-stacktrace.el:
```diff
@@ -41,6 +41,7 @@
 
 (defcustom cider-stacktrace-fill-column t
   "Fill column for error messages in stacktrace display.
+
```
There's no need to add this blank line. The same goes to the 3 below (lines 244, 308, and 485).

turbopape commented:
I eliminated the blank lines and added the space between the [FIX ISSUE] and the rest of the commit message. Any thoughts @bbatsov @Malabarba @expez ?

Malabarba commented:
Looks ok.

turbopape commented:
I updated the Changelog and squashed it into my PR's commit. But now, it shows me conflicts. I refetched from upstream and re-submitted, the conflicts are stille ther. Any idea how I can fix this @bbatsov @Malabarba @expez ?

bbatsov commented:
Looks ok. You'll have to rebase it against the current `master`. You should also change `Added` to `Add` in the message.

bbatsov commented:
This doesn't look right. I see a ton of extra commits in the PR.

turbopape commented:
Yeah. I am struggling with the rebasing with master...

turbopape commented:
I got lost trying to merge the new commits at master. Think I'm gonna close this PR, re-fork and work on a clean copy.
